pkg/config/config.go

```diff
@@ -9,12 +9,92 @@ import (
 
 	"gopkg.in/yaml.v3"
 
+	"github.com/open-policy-agent/opa/ast"
+
 	rio "github.com/styrainc/regal/internal/io"
 )
 
+const capabilitiesEngineOPA = "opa"
+
 type Config struct {
-	Rules  map[string]Category `json:"rules"            yaml:"rules"`
-	Ignore Ignore              `json:"ignore,omitempty" yaml:"ignore,omitempty"`
+	Rules        map[string]Category `json:"rules"            yaml:"rules"`
+	Ignore       Ignore              `json:"ignore,omitempty" yaml:"ignore,omitempty"`
+	Capabilities ast.Capabilities    `json:"capabilities,omitempty"     yaml:"capabilities,omitempty"`
+}
+
+func (config *Config) UnmarshalYAML(value *yaml.Node) error {
+	var result struct {
+		Rules        map[string]Category `yaml:"rules"`
+		Ignore       Ignore              `yaml:"ignore"`
+		Capabilities struct {
+			From struct {
+				Engine  string `yaml:"engine"`
+				Version string `yaml:"version"`
+				File    string `yaml:"file"`
+			} `yaml:"from"`
+			Plus struct {
+				Builtins []*ast.Builtin `yaml:"builtins"`
+			} `yaml:"plus"`
+			Minus struct {
+				Builtins []struct {
+					Name string `yaml:"name"`
+				} `yaml:"builtins"`
+			} `yaml:"minus"`
+		} `yaml:"capabilities"`
+	}
+
+	if err := value.Decode(&result); err != nil {
+		return fmt.Errorf("unmarshalling config failed %w", err)
+	}
+
+	config.Rules = result.Rules
+	config.Ignore = result.Ignore
+
+	capabilitiesFile := result.Capabilities.From.File
+	capabilitiesEngine := result.Capabilities.From.Engine
+	capabilitiesEngineVersion := result.Capabilities.From.Version
+
+	if capabilitiesFile != "" && capabilitiesEngine != "" {
+		return fmt.Errorf("capabilities from.file and from.engine are mutually exclusive")
+	}
+
+	if capabilitiesEngine != "" && capabilitiesEngineVersion == "" {
+		return fmt.Errorf("please set the version for the engine from which to load capabilities from")
+	}
+
+	if capabilitiesFile != "" {
+		bs, err := os.ReadFile(capabilitiesFile)
+		if err != nil {
+			return fmt.Errorf("failed to load capabilities file: %w", err)
+		}
+
+		err = json.Unmarshal(bs, &config.Capabilities)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal capabilities file contents: %w", err)
+		}
+	}
+
+	if capabilitiesEngine != "" && result.Capabilities.From.Engine == capabilitiesEngineOPA {
+		capabilities, err := ast.LoadCapabilitiesVersion(result.Capabilities.From.Version)
+		if err != nil {
+			return fmt.Errorf("loading capabilities failed: %w", err)
+		}
+
+		config.Capabilities = *capabilities
+	}
+
+	// remove any builtins referenced in the minus config
+	for i, builtin := range config.Capabilities.Builtins {
+		for _, minusBuiltin := range result.Capabilities.Minus.Builtins {
+			if minusBuiltin.Name == builtin.Name {
+				config.Capabilities.Builtins = append(config.Capabilities.Builtins[:i], config.Capabilities.Builtins[i+1:]...)
+			}
+		}
+	}
+
+	config.Capabilities.Builtins = append(config.Capabilities.Builtins, result.Capabilities.Plus.Builtins...)
+
+	return nil
```

Comment:
If no capabilities have been provided, should we default to latest from OPA here?

Reply:
I've had a go a finishing this off a little more in 1fff993

Reply:
Awesome! Thanks 😃 Looks good to me.

The one thing I'm not sure about is passing the capabilities from the config to Rego's eval args. Would that not have the linter (i.e. Regal) itself be limited to whatever capabilities the target policy has? I mean, even if we're linting a Rego file with e.g. a pre-`contains` OPA version, surely _we_ should be able to use contains in Regal's linter policy? 

I think the way we'll want to use the capabilities at least initially, is basically as a datasource for what built-in functions we might encounter in the policy we're linting, and have linter rules be able to declare their dependencies... like how we should skip the [custom-has-key-construct](https://docs.styra.com/regal/rules/idiomatic/custom-has-key-construct) if `object.keys` is not in the capabilities, and so on. 

But if we eval with the provided capabilities, it would mean _we_ can't use `object.keys`, no?

Reply:
Ahh, I see. I think I got the wrong end of the stick! I can take this part out.

Reply:
I've taken out the changes in https://github.com/StyraInc/regal/pull/423/commits/5d79a7be38680ff08fa539d8593166ec274c4d4a and then added the defaulting back in https://github.com/StyraInc/regal/pull/423/commits/d292b58d6a3ab12ff136474dbef5f8e28ace0fb8

Reply:
Thanks!